### PR TITLE
Unhardcode mod package loaders

### PR DESCRIPTION
**Pull request by pchote**
This updates #13171 with architectural fixes and addresses the other FileSystem hacks that were blocked on having mod-defined package loaders.

This breaks the cycling dependency between assemblies that define package loaders that contain assemblies that define package loaders that... by requiring all dlls to be loaded from a normal directory.  This then also fixes @abcdefg30's SharpDevelop problems.

This is already a monster of a PR, so i'm going to defer moving all the package loaders to the mod assemblies to a future PR.  I've done the D2K sound parser as a testcase, but moving any more would make it difficult to keep a clean git history while addressing any review comments.

The final commit shows that it is still possible to explicitly mount subdirectories inside zip files.

I strongly suggest reviewing this commit by commit, using the `?w=1` parameter to squash whitespace changes on the main package loader commits.

### CODE REVIEW

**Review by abcdefg30 — APPROVED** [review]
Works, thanks! I can't review the code 100%, but it looked okay as well.

File: OpenRA.Game/FileSystem/ZipFile.cs
Comment (reaperrr) [inline OpenRA.Game/FileSystem/ZipFile.cs]: I thought we dropped .oramod support, is it intentional that you keep support for that extension?

Comment (GraionDilach) [inline OpenRA.Game/FileSystem/ZipFile.cs]: This creeped back from the original PR I believe.

Comment (pchote) [inline OpenRA.Game/FileSystem/ZipFile.cs]: Good catch, fixed.

**Comment by pchote** [comment]
Fixed, rebased, and dropped the test commit.

**Comment by reaperrr** [comment]
> and dropped the test commit.

That was a bit early, could you push a separate test branch? Because I can't get it to work locally, but I'm probably just making a mistake somewhere.

**Comment by reaperrr** [comment]
I can get `ra|bits.zip` to work but not `ra|bits.zip/desert`, *neither on bleed nor here*, though.

**Comment by pchote** [comment]
You need to do `ra|bits.zip: bits` and then `bits|desert`.

**Comment by pchote** [comment]
> That was a bit early, could you push a separate test branch? 

https://github.com/pchote/OpenRA/commit/26eddcfc74052d720c732008f774fbc350ff9ca7

**Comment by reaperrr** [comment]
Test case works 👍 

I'd still prefer either @RoosterDragon, @atlimit8 or someone else who is better than me at reviewing code to look over it before this is merged, so I won't merge this right away.

**Comment by reaperrr** [comment]
I gave the code another look since nobody else seems to have the time or motivation, couldn't find anything that looks wrong, and in addition to the testcase all mods still seem to work fine. I assume that if there is any bug lurking that we overlooked, we'll probably run into it before the next major release, so I'll merge this now.

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c2b9f419e339c9480db65a579d50aee6bdf2ff7c...9c8564a6cf4bc9d2ce209770a24b34ee99bd8b09

**Pull request merged**
